src/rust.rs

```diff
@@ -1257,3 +1258,73 @@ macro_rules! new_jsjitinfo_bitfield_1 {
             (($slotIndex as u32) << 22u32)
     }
 }
+
+pub struct CapturedJSStack<'a> {
+    cx: *mut JSContext,
+    stack: RootedGuard<'a, *mut JSObject>,
+}
+
+impl<'a> CapturedJSStack<'a> {
+    pub fn new(cx: *mut JSContext, mut guard: RootedGuard<'a, *mut JSObject>, max_frame_count: u32) -> Option<Self> { // or maybe another name, for example "capture"?
+        let obj_handle = guard.handle_mut(); // if nullptr, return None
+                                             // how to check if handle is not nullptr?
+                                             // how to handle such errors in a nice way?
+        unsafe {
+            if !CaptureCurrentStack(cx, obj_handle, max_frame_count) {
+                return None;
+            }
+        }
+
+        Some(CapturedJSStack {
+            cx: cx,
+            stack: guard,
+        })
+    }
+
+    pub fn as_string(&self, indent: usize) -> Option<String> {
+        // TODO where errors can occur? some example of non trivial error handling?
```

Comment:
We should verify that [IsSavedFrame](https://github.com/servo/mozjs/blob/3f38435122538e4b66f88878302f81331c0b6a44/mozjs/js/src/jsapi.h#L6039) returns true.

Reply:
Isn't it guaranted by [CaptureCurrentStack](https://dxr.mozilla.org/mozilla-central/source/js/src/jsapi.cpp?q=CaptureCurrentStack&redirect_type=direct#7700) after returning `true`?

Reply:
Yes, but if a consumer does not use the macros and overwrites the value in the `Rooted<*mut JSObject>` before calling as_string, it won't work. This is probably an instance where we would get a `false` value from the code in as_string.